test/expr/closure/closures.swift

```diff
@@ -606,3 +606,86 @@ func testSR14678_Optional() -> (Int, Int)? {
      (print("hello"), 0)
   }
 }
+
+// SR-13239
+func callit<T>(_ f: () -> T) -> T {
+  f()
+}
+
+func callitArgs<T>(_ : Int, _ f: () -> T) -> T {
+  f()
+}
+
+func callitArgsFn<T>(_ : Int, _ f: () -> () -> T) -> T {
+  f()()
+}
+
+func callitGenericArg<T>(_ a: T, _ f: () -> T) -> T { 
+  f()
+}
+
+func callitTuple<T>(_ : Int, _ f: () -> (T, Int)) -> T {
+  f().0
+}
+
+func callitVariadic<T>(_ fs: () -> T...) -> T {
+  fs.first!()
+}
+
+func testSR13239_Tuple() -> Int {
+  // expected-error@+1{{conflicting inferred types from call result and closure argument to generic parameter 'T' ('()' vs. 'Int')}}
+  callitTuple(1) {
+    (print("hello"), 0) 
+  }
+}
+
+func testSR13239() -> Int {
+  // expected-error@+1{{conflicting inferred types from call result and closure argument to generic parameter 'T' ('()' vs. 'Int')}}
+  callit {
+    print("hello")
+  }
+}
+
+func testSR13239_Args() -> Int {
+  // expected-error@+1{{conflicting inferred types from call result and closure argument to generic parameter 'T' ('()' vs. 'Int')}}
+  callitArgs(1) {
+    print("hello") 
+  }
+}
+
+func testSR13239_ArgsFn() -> Int {
+  // expected-error@+1{{conflicting inferred types from call result and closure argument to generic parameter 'T' ('()' vs. 'Int')}}
+  callitArgsFn(1) {
+    { print("hello") } 
+  }
+}
+
+func testSR13239MultiExpr() -> Int {
+  callit {
+    print("hello") 
+    return print("hello") // expected-error {{cannot convert return expression of type '()' to return type 'Int'}}
+  }
+}
+
+func testSR13239_GenericArg() -> Int {
+  // Generic argument is inferred as Int from first argument literal, so no conflict in this case.
+  callitGenericArg(1) {
+    print("hello") // expected-error {{cannot convert value of type '()' to closure result type 'Int'}}
+  }
+}
+
+func testSR13239_Variadic() -> Int {
+  // expected-error@+1{{conflicting inferred types from call result and closure argument to generic parameter 'T' ('()' vs. 'Int')}}
```

Comment:
Sorry, I'm having trouble parsing the meaning of this diagnostic. What do "call result" and "closure argument" mean in this context, and what is a "closure argument to [a] generic parameter"?

Reply:
Conflicting types are inferred for generic parameter from a closure argument and call result positions. Please share if you have a better wording for this!

Reply:
I can understand your explanation better than the diagnostic ("conflicting types inferred for generic parameter 'T'" is a good start, I think), but i'm still not sure which is which: what is a "call result position" and how is it different here from the "closure argument"? Do _TSPL_ or other diagnostic messages refer to these by other more familiar names?

Reply:
I guess we could use existing `conflicting_arguments_for_generic_parameter` but it's also not great because it doesn't make it clear where arguments to `T` are coming from... If you look at the examples it's always closure result type vs. result type of a call where closure was used and it's not necessary that it "expects" certain type but it's contextually inferred for the call too...

Reply:
>  "conflicting types inferred for generic parameter 'T'; closure returns value of type 'Void' but outer function expects return value of type 'Int'" (but less wordy)?

It is also not always a closure return directly, but closure result involving the generic parameter in some way e.g. tuple and function in the test cases...

> I guess we could use existing conflicting_arguments_for_generic_parameter but it's also not great because it doesn't make it clear where arguments to T are coming from...

Yeah, agree

Reply:
Don't know if is actually better, but here is an idea `conflicting types for generic parameter 'T'; ambiguous inference from 'fuctionName' result and closure argument #0 contexts ('()' vs. 'Int')`. WDYT?

Reply:
> If you look at the examples it's always closure result type vs. result type of a call where closure was used

Then we should say exactly that: "conflicting types inferred for generic parameter 'T' from closure return value and ____". I'm not sure about the blank there because the closure isn't _used_ in the sense of being invoked.

Reply:
How about this - we use existing error diagnostic for conflicting arguments and add two notes  - one for closure e.g. `generic parameter 'T' inferred as 'Int' from closure return expression` and another one for the call - `generic parameter 'T' inferred as 'Void' from context'?

Reply:
Sure, that would make sense to me!

Reply:
Fixed =]

Reply:
Hey @xedin 
Is any other adjustment to make? 
Also, It seems that there is something up with macOS CI ... running it again to see

Reply:
Looks good to me! Hold off on triggering macOS CI there currently as issue with it which is being investigated.

Reply:
> Hold off on triggering macOS CI there currently as issue with it which is being investigated

Ah sure, thanks for the info :))